hudi-common/src/main/java/org/apache/hudi/common/table/timeline/dto/FileGroupDTO.java

```diff
@@ -46,17 +49,50 @@ public class FileGroupDTO {
   TimelineDTO timeline;
 
   public static FileGroupDTO fromFileGroup(HoodieFileGroup fileGroup) {
+    return fromFileGroup(fileGroup, true);
+  }
+
+  public static List<FileGroupDTO> fromFileGroup(List<HoodieFileGroup> fileGroups) {
+    if (fileGroups.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    List<FileGroupDTO> fileGroupDTOs = fileGroups.stream()
+        .map(fg -> FileGroupDTO.fromFileGroup(fg, false)).collect(Collectors.toList());
+    // Timeline exists only in the first file group DTO. Optimisation to reduce payload size.
+    fileGroupDTOs.set(0, FileGroupDTO.fromFileGroup(fileGroups.get(0), true));
+    return fileGroupDTOs;
+  }
+
+  public static HoodieFileGroup toFileGroup(FileGroupDTO dto, HoodieTableMetaClient metaClient) {
+    return toFileGroup(dto, metaClient, null);
+  }
+
+  public static Stream<HoodieFileGroup> toFileGroup(List<FileGroupDTO> dtos, HoodieTableMetaClient metaClient) {
+    if (dtos.isEmpty()) {
+      return Stream.empty();
+    }
+
+    // Timeline exists only in the first file group DTO. Optimisation to reduce payload size.
+    HoodieTimeline timeline = toFileGroup(dtos.get(0), metaClient).getTimeline();
+    return dtos.stream().map(dto -> toFileGroup(dto, metaClient, timeline));
```

Comment:
Please add validation for the code path: `ValidationUtils.checkState` to ensure the existence of the timeline.